python/cuml/_thirdparty/sklearn/preprocessing/_imputation.py

```diff
@@ -379,27 +379,40 @@ def _dense_fit(self, X, strategy, missing_values, fill_value):
             count_missing_values = mask.sum(axis=0)
             n_elems = X.shape[0] - count_missing_values
             mean = np.nansum(X, axis=0)
-            mean -= (count_missing_values * missing_values)
+            if not np.isnan(missing_values):
+                mean -= (count_missing_values * missing_values)
             mean /= n_elems
             return mean
 
         # Median
         elif strategy == "median":
+            if X.size == 0:
+                return np.full(X.shape[1], np.nan)
+            X_sorted = X.copy()
+            if not np.isnan(missing_values):
+                # If nan is not the missing value, any column with nans should
+                # have a median of nan
+                nan_cols = np.any(np.isnan(X), axis=0)
```

Comment:
Is all of this because there is no cp.median?

Reply:
Just a quick note, cupy added median in 8.0 and the PR for nanmedian just landed two days ago https://github.com/cupy/cupy/pull/4092

Reply:
Yes, once we switch to the next version of cupy, we should move most of this logic over to use cp.nanmedian. We can use a similar trick to what we did here for non-nan cases but clean up the code quite a bit.